### PR TITLE
Handle SIGABRT in the client application elegantly #27

### PR DESCRIPTION
**Pull request by ahmadsyafrudin**
Detail of this Pull Request :
https://github.com/t0xic0der/sanctuary-zero/issues/27#issuecomment-708610376

### CODE REVIEW

**Comment by t0xic0der** [comment]
There's one self-review I would ask you to perform @ahmadsyafrudin. Could you please enter the key combination `Ctrl+D` in the client application and in the server application, and see what happens. Do send here the resulting screenshots as well.

**Review by t0xic0der — CHANGES_REQUESTED** [review]
Hi, I reviewed your pull request and found a couple points that need your attention. 

- This is me pressing `Ctrl+D` to send an abort signal. As expected, your code helps to elegantly exit the client but does not address the broken connection at the server end - which still believes that the client is connected. 
image

- This is at the server side. Even after the client has faced an abort signal with the key combination, it does not show that the client has left. As a result, it pushes all the messages to all clients (even the broken ones because the server does not know that the client has disconnected).
image

**Comment by t0xic0der** [comment]
Awesome work with implementing the elegant client exit when the connection has been severed from the server's end.
image

**Comment by ahmadsyafrudin** [comment]
okay working On it for the revision Tonight

**Comment by ahmadsyafrudin** [comment]
> Hi, I reviewed your pull request and found a couple points that need your attention.
> 
> * This is me pressing `Ctrl+D` to send an abort signal. As expected, your code helps to elegantly exit the client but does not address the broken connection at the server end - which still believes that the client is connected.
>   image
> * This is at the server side. Even after the client has faced an abort signal with the key combination, it does not show that the client has left. As a result, it pushes all the messages to all clients (even the broken ones because the server does not know that the client has disconnected).
>   image

@t0xic0der , yes. Already figure it out. Ctrl+D Is not signal like Ctrl+C. its work on client because we use input in there. Ctrl+D is for end of file: `Use raw_input (Use input in Python 3.x). Pressing Ctrl+D will cause EOFError exception.`. will give the Update for the #32

**Comment by t0xic0der** [comment]
Umm. That makes sense. Please make the necessary changes so that we get the expected behaviour.

**Comment by t0xic0der** [comment]
Hi @ahmadsyafrudin, any progress you made so far? :sweat_smile:

**Comment by ahmadsyafrudin** [comment]
Screen Recording 2020-10-17 at 04 28 24

I just play little bit with prompt toolkit. I already wrote Todo if you want to make UI refreshed properly. 
I hope it's still OK since all functionality are skookum.
Thanks

**Comment by ahmadsyafrudin** [comment]
solved case :
- Gracefully closed connection when client press Ctrl+d (Server received message that client are exited)
- Gracefully closed connection when client press Ctrl+d on server.(Added input Context so server can detect Ctrl+d)

**Comment by t0xic0der** [comment]
> hi @t0xic0der , sorry for late response. haha. I'll give the update soon. Thanks for reach me out 😬 . don't worry sir

No worries. Glad to know that you were still working on the issue. :+1:

**Comment by t0xic0der** [comment]
I'll get to review this PR right away.

File: cnew.py [hunk 1]
```diff
@@ -44,24 +46,40 @@ async def consumer_handler(cphrsuit, websocket, username, chatroom, servaddr):
...
+        while True:
+            with patch_stdout():
+                mesgtext = await sess.prompt_async(lambda:"[" + obtntime() + "] " + formusnm(str(username)) + " > ", bottom_toolbar=footelem, validator=emtyfind(), refresh_interval=0.5)
+            senddata = json.dumps({"username": username.strip(), "chatroom": chatroom, "mesgtext": mesgtext.strip()})
+            senddata = cphrsuit.encrjson(senddata)
+            await websocket.send(senddata)
+
+    except EOFError:
+        print("Quit, Ctrl+D pressed")
+        raise KeyboardInterrupt
```
Comment (t0xic0der) [inline cnew.py:60]: Excellent work done here! The only thing that I would as you to change is to get rid of the `print("Quit, Ctrl+D pressed")` statement at L#59.

File: cnew.py [hunk 2]
```diff
@@ -44,24 +46,40 @@ async def consumer_handler(cphrsuit, websocket, username, chatroom, servaddr):
...
-        prod = asyncio.get_event_loop().create_task(producer_handler(cphrsuit, websocket, str(username), str(chatroom), str(servaddr)))
-        cons = asyncio.get_event_loop().create_task(consumer_handler(cphrsuit, websocket, str(username), str(chatroom), str(servaddr)))
-        await websocket.send(username+sepr+chatroom)
-        await prod
-        await cons
-        asyncio.get_event_loop().run_forever()
+        try:
+            cphrsuit = fernetst(password.encode("utf8"))
+            prod = asyncio.get_event_loop().create_task(
+                producer_handler(cphrsuit, websocket, str(username), str(chatroom), str(servaddr)))
```
Comment (t0xic0der) [inline cnew.py:68]: Please restrict the statement to a single line.

File: cnew.py [hunk 3]
```diff
@@ -44,24 +46,40 @@ async def consumer_handler(cphrsuit, websocket, username, chatroom, servaddr):
...
+            cons = asyncio.get_event_loop().create_task(
+                consumer_handler(cphrsuit, websocket, str(username), str(chatroom), str(servaddr)))
```
Comment (t0xic0der) [inline cnew.py:70]: Please restrict the statement to a single line.

File: cnew.py [hunk 4]
```diff
@@ -44,24 +46,40 @@ async def consumer_handler(cphrsuit, websocket, username, chatroom, servaddr):
...
+            await cons
+            asyncio.get_event_loop().run_forever()
+        except Exception as e:
+            if websocket.closed:
+                print_formatted_text(
+                    HTML(
+                        "[" + obtntime() + "] "
+                        + "SNCTRYZERO > <red>Can't Establish Connection to server. Detail: {}</red>".format(e))
+                )
+            raise KeyboardInterrupt
```
Comment (t0xic0der) [inline cnew.py:82]: Nice work with detecting the connection closure from the server end. :+1: The only thing I would ask here is to restrict the L#77:81 statements to a single line.

File: cnew.py [hunk 5]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
 
 
 def formusnm(username):
-    if len(username) < 10:      return username + " " * (10 - len(username))
-    elif len(username) > 10:    return username[0:10]
-    else:                       return username
+    if len(username) < 10:
+        return username + " " * (10 - len(username))
```
Comment (t0xic0der) [inline cnew.py:123]: L#122:123 were previously on the same line. Please restore the same.

File: cnew.py [hunk 6]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
+    elif len(username) > 10:
+        return username[0:10]
```
Comment (t0xic0der) [inline cnew.py:125]: L#124:125 were previously on the same line. Please restore the same.

File: cnew.py [hunk 7]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
+    else:
+        return username
```
Comment (t0xic0der) [inline cnew.py:127]: L#126:127 were previously on the same line. Please restore the same.

File: cnew.py [hunk 8]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
 
 
 @click.command()
-@click.option("-u", "--username", "username", help="Enter the username that you would identify yourself with", required=True)
+@click.option("-u", "--username", "username", help="Enter the username that you would identify yourself with",
+              required=True)
```
Comment (t0xic0der) [inline cnew.py:132]: L#131:132 were previously on the same line. Please restore the same.

File: cnew.py [hunk 9]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
 @click.option("-p", "--password", "password", help="Enter the chatroom password for decrypting the messages")
 @click.option("-c", "--chatroom", "chatroom", help="Enter the chatroom identity you would want to enter in")
-@click.option("-s", "--servaddr", "servaddr", help="Enter the server address you would want to connect to", required=True)
+@click.option("-s", "--servaddr", "servaddr", help="Enter the server address you would want to connect to",
+              required=True)
```
Comment (t0xic0der) [inline cnew.py:136]: L#135:136 were previously on the same line. Please restore the same.

File: cnew.py [hunk 10]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
 @click.version_option(version="04092020", prog_name="SNCTRYZERO Client by t0xic0der")
 def mainfunc(username, password, chatroom, servaddr):
     try:
         os.system("clear")
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <b><seagreen>Starting Sanctuary ZERO v04092020 up...</seagreen></b>"))
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Connected to '" + servaddr + "' successfully</seagreen>"))
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Session started at " + str(time.ctime()) + "</seagreen>"))
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <b><seagreen>Starting Sanctuary ZERO v04092020 up...</seagreen></b>"))
```
Comment (t0xic0der) [inline cnew.py:142]: L#141:142 were previously on the same line. Please restore the same.

File: cnew.py [hunk 11]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Connected to '" + servaddr + "' successfully</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:144]: L#143:144 were previously on the same line. Please restore the same.

File: cnew.py [hunk 12]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Session started at " + str(time.ctime()) + "</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:146]: L#145:146 were previously on the same line. Please restore the same.

File: cnew.py [hunk 13]
```diff
@@ -101,25 +119,34 @@ def chekpass(pswd):
...
         if chatroom is None:
-            print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <green>A new chatroom was generated</green>"))
+            print_formatted_text(
+                HTML("[" + obtntime() + "] " + "SNCTRYZERO > <green>A new chatroom was generated</green>"))
```
Comment (t0xic0der) [inline cnew.py:149]: L#148:149 were previously on the same line. Please restore the same.

File: cnew.py [hunk 14]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
             else:
                 print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>An invalid chatroom password was entered</red>"))
                 sys.exit()
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Chatroom identity : " + chatroom + "</seagreen>"))
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Chatroom password : " + password + "</seagreen>"))
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Share the chatroom identity and password to add members!</seagreen>"))
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Your conversations are protected with end-to-end encryption</seagreen>"))
+        print_formatted_text(
+            HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Chatroom identity : " + chatroom + "</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:174]: L#173:174 were previously on the same line. Please restore the same.

File: cnew.py [hunk 15]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
+        print_formatted_text(
+            HTML("[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Chatroom password : " + password + "</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:176]: L#175:176 were previously on the same line. Please restore the same.

File: cnew.py [hunk 16]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Share the chatroom identity and password to add members!</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:178]: L#177:178 were previously on the same line. Please restore the same.

File: cnew.py [hunk 17]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <seagreen>Your conversations are protected with end-to-end encryption</seagreen>"))
```
Comment (t0xic0der) [inline cnew.py:180]: L#179:180 were previously on the same line. Please restore the same.

File: cnew.py [hunk 18]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
         asyncio.get_event_loop().run_until_complete(hello(servaddr, username, chatroom, password))
     except KeyboardInterrupt as EXPT:
         print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>Leaving SNCTRYZERO...</red>"))
         sys.exit()
     except OSError as EXPT:
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>A connection to the server could not be established</red>"))
+        print_formatted_text(HTML(
+            "[" + obtntime() + "] " + "SNCTRYZERO > <red>A connection to the server could not be established</red>"))
```
Comment (t0xic0der) [inline cnew.py:187]: L#186:187 were previously on the same line. Please restore the same.

File: cnew.py [hunk 19]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
         sys.exit()
     except websockets.exceptions.ConnectionClosedError as EXPT:
-        print_formatted_text(HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>A connection to the server was lost</red>"))
+        print_formatted_text(
+            HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>A connection to the server was lost</red>"))
```
Comment (t0xic0der) [inline cnew.py:191]: L#190:191 were previously on the same line. Please restore the same.

File: cnew.py [hunk 20]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
         sys.exit()
 
 
+def sig_handler(signum, frame):
+    print_formatted_text(
+        HTML("[" + obtntime() + "] " + "SNCTRYZERO > <red>SIGABRT{} is being called</red>".format(signum)))
```
Comment (t0xic0der) [inline cnew.py:197]: L#196:197 were previously on the same line. Please restore the same.

File: cnew.py [hunk 21]
```diff
@@ -143,21 +170,35 @@ def mainfunc(username, password, chatroom, servaddr):
...
+    raise KeyboardInterrupt
+
+
+signal.signal(signal.SIGABRT, sig_handler)
```
Comment (t0xic0der) [inline cnew.py:201]: I don't think we would be needing this anymore. Are there ways to trigger a SIGABRT from inside the application?

**Review by t0xic0der — CHANGES_REQUESTED** [review]
Follow the comments provided for the `cnew.py`.

image

I don't think we would be needing a full-screen CLI for the server application. You are requested to restore the previous look and feel of the server application while retaining the actions you wrote that deal with the keybindings.

**Comment by t0xic0der** [comment]
And yes, there are extra newlines too that you would want to remove.

**Comment by ahmadsyafrudin** [comment]
Will give the update with requested Code revie comment by tonight. Thanks 👍

**Comment by ahmadsyafrudin** [comment]
hi @t0xic0der , Since the changes are minor. I Updated the Pull request just right now. 😄

**Comment by t0xic0der** [comment]
Did you consider the suggestions made here https://github.com/t0xic0der/sanctuary-zero/pull/32#pullrequestreview-510953685?

File: cnew.py (hunk 21 above)
Comment (ahmadsyafrudin) [reply cnew.py]: yes, if client app is being killed, it's still sending Exited information to server. 
Screen Shot 2020-10-17 at 13 45 38

**Comment by ahmadsyafrudin** [comment]
> Did you consider the suggestions made here [#32 (review)](https://github.com/t0xic0der/sanctuary-zero/pull/32#pullrequestreview-510953685)?

Hi @t0xic0der ,
I Also Restore Previous Look And Feel. but input context is still there:
sanctuary

**Comment by t0xic0der** [comment]
> > Did you consider the suggestions made here [#32 (review)](https://github.com/t0xic0der/sanctuary-zero/pull/32#pullrequestreview-510953685)?
> 
> Hi @t0xic0der ,
> I Also Restore Previous Look And Feel. but input context is still there:
> sanctuary

 Cool. Let me get into reviewing it.

**Comment by t0xic0der** [comment]
@ahmadsyafrudin `Ctrl+D` and `Ctrl+Q` actions are inconsequential to the server app as no input is taken there. So you could rollback the input contexts as well.

**Comment by ahmadsyafrudin** [comment]
> `Ctrl+D` and `Ctrl+Q` actions are inconsequential to the server app as no input is taken there. So you could rollback the input contexts as well.

So we don't need handle CTRL+D and CTRL+Q ?

**Comment by t0xic0der** [comment]
> > `Ctrl+D` and `Ctrl+Q` actions are inconsequential to the server app as no input is taken there. So you could rollback the input contexts as well.
> 
> @t0xic0der So we don't need to handle CTRL+D and CTRL+Q ?

Yes. We don't need to handle CTRL+D and CTRL+Q at the server-side. We still need them on the client-side though.

**Comment by ahmadsyafrudin** [comment]
still need the signal handler on server side? @t0xic0der

**Comment by ahmadsyafrudin** [comment]
Hi, @t0xic0der . Input context already removed. and also the SIGABRT Handler are removed.

**Comment by t0xic0der** [comment]
Cool. Reviewing now.

File: cnew.py [hunk 22]
```diff
@@ -1,9 +1,11 @@
 import asyncio, websockets, time, json, click, secrets, os, sys
+
+
 from prompt_toolkit import PromptSession
 from prompt_toolkit.patch_stdout import patch_stdout
 from prompt_toolkit import print_formatted_text, HTML
 from prompt_toolkit.validation import Validator, ValidationError
-from cryptography.fernet import Fernet, InvalidToken
+from cryptography.fernet import Fernet
```
Comment (t0xic0der) [inline cnew.py:8]: Thank you for restoring the import styles.

**Review by t0xic0der — APPROVED** [review]
Tested it locally. `Ctrl+D` is correctly answered to at the client-side and the connection closure from the server is correctly dealt with. 


image

**Pull request merged**